File: backend/app/legal.py
"""Валидация данных регистрации юридического лица (отдельный модуль).

Не связан с логикой каталога вакансий: телефон (маска
+7 (XXX) XXX-XX-XX), ИНН юрлица (10 цифр + контрольная сумма)
и пароль (не менее 8 символов, латиница в верхнем и нижнем
регистре, цифры).
"""

import re

# Веса контрольной суммы ИНН юрлица (первые 9 разрядов).
INN_WEIGHTS = (2, 4, 10, 3, 5, 9, 4, 6, 8)

_NON_DIGITS = re.compile(r"\D+")


def normalize_phone(phone: str) -> str | None:
    """Приводит телефон к каноническому виду +7XXXXXXXXXX.

    Принимает любые разделители (пробелы, скобки, дефисы) и варианты
    префикса (7, 8 или вообще без префикса). Возвращает None, если
    номер не похож на российский 11-значный.
    """
    digits = _NON_DIGITS.sub("", phone)
    if len(digits) == 11 and digits[0] in ("7", "8"):
        return f"+7{digits[1:]}"
    if len(digits) == 10:
        return f"+7{digits}"
    return None


def is_valid_inn_10(inn: str) -> bool:
    """Проверяет ИНН юрлица: ровно 10 цифр и контрольная сумма."""
    if len(inn) != 10 or not inn.isdigit():
        return False
    control = sum(int(inn[i]) * INN_WEIGHTS[i] for i in range(9)) % 11 % 10
    return control == int(inn[9])


def is_valid_password(password: str) -> bool:
    """Проверяет пароль: не менее 8 символов, латинские буквы
    верхнего и нижнего регистра и хотя бы одна цифра."""
    if len(password) < 8:
        return False
    if not re.search(r"[a-z]", password):
        return False
    if not re.search(r"[A-Z]", password):
        return False
    if not re.search(r"[0-9]", password):
        return False
    return True

File: frontend/src/lib/legal.js
/* Утилиты формы регистрации юридического лица (отдельный модуль —
   не пересекается с логикой каталога вакансий: lib/query.js, lib/format.js). */

/* Веса контрольной суммы ИНН юрлица (первые 9 разрядов). */
const INN_WEIGHTS = [2, 4, 10, 3, 5, 9, 4, 6, 8];

/* Проверка пароля: не менее 8 символов, латинские буквы верхнего
   и нижнего регистра и хотя бы одна цифра. */
export function isValidPassword(password) {
  if (password.length < 8) return false;
  if (!/[a-z]/.test(password)) return false;
  if (!/[A-Z]/.test(password)) return false;
  if (!/[0-9]/.test(password)) return false;
  return true;
}

/* Проверка ИНН юрлица: ровно 10 цифр + контрольная сумма. */
export function isValidInn10(inn) {
  if (!/^\d{10}$/.test(inn)) return false;
  let sum = 0;
  for (let i = 0; i < 9; i += 1) sum += Number(inn[i]) * INN_WEIGHTS[i];
  return sum % 11 % 10 === Number(inn[9]);
}

/* Маска телефона +7 (XXX) XXX-XX-XX.
   На вход — только цифры номера без префикса (до 10 цифр).
   Пример: "9123456789" → "+7 (912) 345-67-89".
   Пустой номер → "", чтобы поле можно было полностью очистить
   (плейсхолдер показывает маску). */
export function formatPhoneDigits(digits) {
  if (digits.length === 0) return '';
  let out = '+7';
  if (digits.length > 0) out += ` (${digits.slice(0, 3)}`;
  if (digits.length >= 3) out += ')';
  if (digits.length > 3) out += ` ${digits.slice(3, 6)}`;
  if (digits.length > 6) out += `-${digits.slice(6, 8)}`;
  if (digits.length > 8) out += `-${digits.slice(8, 10)}`;
  return out;
}

/* Извлекает цифры номера из введённого значения поля с маской.

   В поле всегда присутствует префикс "+7": при удалении фрагмента
   или правке середины номера цифра префикса "7" попадает в выборку
   и портит номер — поэтому, если строка начинается с "+", первая
   цифра (префикс) отбрасывается. Вставка целиком без "+"
   ("7912…"/"8912…", 11 цифр) тоже сбрасывает первую цифру. */
export function digitsFromPhoneRaw(value) {
  let digits = value.replace(/\D+/g, '');
  if (value.trimStart().startsWith('+') && digits.length > 0 && (digits[0] === '7' || digits[0] === '8')) {
    digits = digits.slice(1);
  } else if (digits.length === 11 && (digits[0] === '7' || digits[0] === '8')) {
    digits = digits.slice(1);
  }
  return digits.slice(0, 10);
}
